src/formatting.rs

```diff
@@ -683,6 +701,45 @@ fn parse_crate(
     Err(ErrorKind::ParseError)
 }
 
+struct SilentOnIgnoredFilesEmitter {
+    ignore_path_set: Rc<IgnorePathSet>,
+    source_map: Rc<SourceMap>,
+    emitter: EmitterWriter,
+    can_reset: bool,
```

Comment:
Sure! I was trying to avoid having to do a `borrow_mut` and update the value unless absolutely necessary, and that was accomplished via both `can_reset` and `has_non_ignorable_parser_errors`. If there was only `has_non_ignorable_parser_errors`, then there could be a lot of repetitive `borrow_mut` setting the value to `true` over and over. 

I'm not sure if/how problematic that would be, but that was my thought process anyway. I could add some inline comments to explain that, or I could remove it altogether if you'd prefer

Reply:
Ah, so is `can_reset` used to avoid the overhead of calling `borrow_mut` multiple times? If that is the case, IMHO, this seems premature optimization to me. `RefMut::borrow_mut` is quite cheap (some non-atomic counter operations), and the code path is not hot anyway (it's only run on parse error).

Reply:
Is it possible to remove `can_reset` and `has_non_ignorable_parser_errors` without changing the semantic? If possible, then I prefer to remove them to keep the code simple.

Reply:
> Ah, so is can_reset used to avoid the overhead of calling borrow_mut multiple times?

Basically yes, and also because I wasn't sure if there'd ever be any competing borrow attempts which I'd [read](https://doc.rust-lang.org/beta/std/cell/struct.RefCell.html#method.borrow_mut) would panic. Sounds like there's no issues there so I'll remove that 👍 

> Is it possible to remove can_reset and has_non_ignorable_parser_errors without changing the semantic?

I can remove one of them, but AFAICT we'll need at least one flag to indicate whether parser errors from _non-ignored_ files were encountered to ensure that we still fail in such a scenario. For example, if there are two files that have parser errors, `foo.rs` and `bar.rs`, and `bar.rs` is ignored, and `foo` is processed first and then `bar`. We should still fail due to the parser error in `foo.rs`

Reply:
I've updated this to only use the one flag. The only way I could see to remove them both would be to set the top level reset value to `true`, and then updating the emitter logic. 
https://github.com/rust-lang/rustfmt/blob/0531683677eb82dc9f586e9ad51ad1a35bd18e29/src/formatting.rs#L79

However, I don't think that the default behavior should be to ignore/reset parse errors encountered while parsing the crate. I fear that could result in spurious ignores of parser errors which would later cause problems when trying to parse the modules.

IMO it's better to have the emitter keep track of whether it saw a parser error from a non-ignored file via a flag, but please let me know if you'd prefer I change to the other approach!